### ISSUE TITLE
How to use newlib to compile c code?

### ISSUE THREAD

**Issue by ShervinShokouhi** [sylefeb/Silice#219]
Assume you have below code (neglect about puts and putchar) how I cane use newlib to compile this code?

Compile cmd:
riscv32-unknown-elf-gcc -fno-unroll-loops -O1 -fno-stack-protector -fno-pic -march=rv32i -mabi=ilp32 -c -o test.o test.c

Code:
#include <stdio.h>
int main()
{
  int k = 2;
  float i,j,r,x,y=-16;
  while (puts(""), y++<15)
    for (x=0; x++<84; putchar(" .:-;!/>)|&IH%*#"[k&15]))
      for (i=k=r=0; 
           j=r*r-i*i-2+x/25,i=2*r*i+y/10,j*j+i*i<11&&k++<111; 
           r=j);
  return 0;
}

I get these error messages:
 riscv32-unknown-elf-ld  -m elf32lriscv -b elf32-littleriscv --no-relax  -lm -o test.elf test.o               F:\Silice\Toolchains\riscv32-unknown-elf\bin\riscv32-unknown-elf-ld.exe: warning: cannot find entry symbol _start; defaulting to 0000000000010074
F:\Silice\Toolchains\riscv32-unknown-elf\bin\riscv32-unknown-elf-ld.exe: test.o: in function `main':
test.c:(.text+0x40): undefined reference to `puts'
F:\Silice\Toolchains\riscv32-unknown-elf\bin\riscv32-unknown-elf-ld.exe: test.c:(.text+0xb4): undefined reference to `putchar'
F:\Silice\Toolchains\riscv32-unknown-elf\bin\riscv32-unknown-elf-ld.exe: test.c:(.text+0xc8): undefined reference to `__addsf3'
F:\Silice\Toolchains\riscv32-unknown-elf\bin\riscv32-unknown-elf-ld.exe: test.c:(.text+0xdc): undefined reference to `__ltsf2'
F:\Silice\Toolchains\riscv32-unknown-elf\bin\riscv32-unknown-elf-ld.exe: test.c:(.text+0xf4): undefined reference to `__divsf3'
F:\Silice\Toolchains\riscv32-unknown-elf\bin\riscv32-unknown-elf-ld.exe: test.c:(.text+0x108): undefined reference to `__divsf3'
F:\Silice\Toolchains\riscv32-unknown-elf\bin\riscv32-unknown-elf-ld.exe: test.c:(.text+0x130): undefined reference to `__mulsf3'
F:\Silice\Toolchains\riscv32-unknown-elf\bin\riscv32-unknown-elf-ld.exe: test.c:(.text+0x144): undefined reference to `__mulsf3'
F:\Silice\Toolchains\riscv32-unknown-elf\bin\riscv32-unknown-elf-ld.exe: test.c:(.text+0x154): undefined reference to `__subsf3'
F:\Silice\Toolchains\riscv32-unknown-elf\bin\riscv32-unknown-elf-ld.exe: test.c:(.text+0x160): undefined reference to `__subsf3'
F:\Silice\Toolchains\riscv32-unknown-elf\bin\riscv32-unknown-elf-ld.exe: test.c:(.text+0x16c): undefined reference to `__addsf3'
F:\Silice\Toolchains\riscv32-unknown-elf\bin\riscv32-unknown-elf-ld.exe: test.c:(.text+0x180): undefined reference to `__addsf3'
F:\Silice\Toolchains\riscv32-unknown-elf\bin\riscv32-unknown-elf-ld.exe: test.c:(.text+0x18c): undefined reference to `__mulsf3'
F:\Silice\Toolchains\riscv32-unknown-elf\bin\riscv32-unknown-elf-ld.exe: test.c:(.text+0x198): undefined reference to `__addsf3'
F:\Silice\Toolchains\riscv32-unknown-elf\bin\riscv32-unknown-elf-ld.exe: test.c:(.text+0x1ac): undefined reference to `__mulsf3'
F:\Silice\Toolchains\riscv32-unknown-elf\bin\riscv32-unknown-elf-ld.exe: test.c:(.text+0x1c0): undefined reference to `__mulsf3'
F:\Silice\Toolchains\riscv32-unknown-elf\bin\riscv32-unknown-elf-ld.exe: test.c:(.text+0x1d0): undefined reference to `__addsf3'
F:\Silice\Toolchains\riscv32-unknown-elf\bin\riscv32-unknown-elf-ld.exe: test.c:(.text+0x1dc): undefined reference to `__ltsf2'
F:\Silice\Toolchains\riscv32-unknown-elf\bin\riscv32-unknown-elf-ld.exe: test.c:(.text+0x1fc): undefined reference to `puts'
F:\Silice\Toolchains\riscv32-unknown-elf\bin\riscv32-unknown-elf-ld.exe: test.c:(.text+0x20c): undefined reference to `__addsf3'
F:\Silice\Toolchains\riscv32-unknown-elf\bin\riscv32-unknown-elf-ld.exe: test.c:(.text+0x220): undefined reference to `__ltsf2'

Help me about GCC, AS, LD and rest settings to create HEX file from C with NewLib support.

**Comment by rob-ng15** [comment]
You'll need to pass a flag to link to newlib and the gcc support libraries.

These will be located at something like:

/usr/riscv64-elf/lib/rv32i/ilp32f/libc.a
/usr/riscv64-elf/lib/rv32i/ilp32f/libm.a
/usr/lib/gcc/riscv64-elf/12.1.0/rv32i/ilp32f/libgcc.a

Don't worry about the riscv64 as these are the 32 bit versions of the
libraries. Yours may indeed be at riscv32.

You'll also need a crt0.o file which provides the _start symbol,
initialises the stack pointer and jumps to main().

So something like:

riscv32-unknown-elf-gcc -fno-unroll-loops -O1 -fno-stack-protector -fno-pic
-march=rv32i -mabi=ilp32  \
/usr/riscv64-elf/lib/rv32i/ilp32f/libc.a \
/usr/riscv64-elf/lib/rv32i/ilp32f/libm.a \
/usr/lib/gcc/riscv64-elf/12.1.0/rv32i/ilp32f/libgcc.a \
-c -o test.o test.c

This doesn't provide crt0.o though, but it should get you further.

Alternatively, use theis compile script from Sylefeb as a basis?

https://github.com/sylefeb/Silice/blob/master/projects/ice-v/compile/icestick/compile_c.sh

**Comment by ShervinShokouhi** [comment]
Thanks for your help, but where is "--specs=nano.specs" inside "https://github.com/sylefeb/Silice/blob/master/projects/ice-v/compile/icestick/compile_c.sh" script?
I used ice-v scripts but caused: undefined reference to __subsf3'

**Comment by rob-ng15** [comment]
Did you include /usr/lib/gcc/riscv64-elf/12.1.0/rv32i/ilp32f/libgcc.a as
that should be in there?

**Comment by rob-ng15** [comment]
Yes, I believe that the floating point emulation is in the libgcc.a file.

**Comment by rob-ng15** [comment]
Just realised that I've given the paths to the versions that have hardware floating point support, I think you need to remove the f from the end of ip32

**Comment by rob-ng15** [comment]
So the correct line is:

riscv32-unknown-elf-gcc -fno-unroll-loops -O1 -fno-stack-protector -fno-pic
-march=rv32i -mabi=ilp32  \
/usr/riscv64-elf/lib/rv32i/ilp32/libc.a \
/usr/riscv64-elf/lib/rv32i/ilp32/libm.a \
/usr/lib/gcc/riscv64-elf/12.1.0/rv32i/ilp32/libgcc.a \
-c -o test.o test.c

That was my error, the correct libraries are listed above. My CPU has
floating point support, so ilp32f is correct, without floating point it is
ilp32.

Try those libraries with the linker script and you should be fine.

Apologies.

**Comment by ShervinShokouhi** [comment]
Hi @sylefeb, @rob-ng15 
Thanks for your detailed descriptions, now I'm able to compile C file to O. But in linker stage again asks for MATH functions:

C:\msys64\mingw64\bin\riscv64-unknown-elf-ld.exe: main.c:(.text+0x1e0): undefined reference to `__mulsf3'

How to link newlib library in linker stage?

**Comment by rob-ng15** [comment]
Right, I didn't know that you were using Ming, I only use gcc. There must be an equivalent for Ming?

**Comment by sylefeb** [comment]
I could not test beyond compilation yet but this seems to work on my side:
https://gist.github.com/sylefeb/64a0a55681f0680908989724c05e1778

Might need a change in paths, for libgcc in particular as it contains a version number (note: the order of object file / libraries does matter in the ld command line).

**Comment by ShervinShokouhi** [comment]
I also compiled it :)
Thank you @sylefeb 
But I have some problem with test it in my custom SOC.
I'll write about results

**Comment by ShervinShokouhi** [comment]
Another question: Now generated HEX file contains several sections split by something like "@000001E0", how do you fill them into BRAM for CPU?

**Comment by sylefeb** [comment]
Hex file: I have updated the preprocessor script to do that: https://github.com/sylefeb/Silice/blob/draft/projects/ice-v/SOCs/pre_include_compiled.lua (could use a rewrite but does the job ;) ).

**Comment by sylefeb** [comment]
> I'll write about results

Please let us know, hopefully we can make it work as well.

Btw, after a bit of searching, there is a simpler approach! The trick is to let gcc do the link because it knows where the libs are (found the hint at the bottom of the answer: https://stackoverflow.com/questions/64603070/risc-v-linker-cannot-find-lgcc ).

We can then use `-lm` (for math) and `-nostartfiles` (to use our custom crt0).

```
$ARCH-as -march=rv32i -mabi=ilp32 -o crt0.o $BASE/crt0.s
$ARCH-gcc -DICESTICK -nostartfiles -fno-builtin -fno-unroll-loops -O1 -fno-stack-protector -fno-pic -march=rv32i -mabi=ilp32 -T $BASE/config_c.ld -o $DST/code.elf $1 crt0.o -lm
```

That is way cleaner, I'll update my scripts accordingly.

**Comment by ShervinShokouhi** [comment]
I compiled my code and tested it.
About your latest compilation script tested, I don't check it yet.
Newlib is fantastic, thank you @sylefeb.

**Comment by sylefeb** [comment]
Thanks for the feedback and tests! I'll close the issue for now but feel free to reopen if needed.
